hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -131,7 +132,11 @@ the following fields are used in all circumstances, and are not worth (or not su
 	private IdentityHashMap<Object, PersistentCollection<?>> arrayHolders;
 
 	// Identity map of CollectionEntry instances, by the collection wrapper
-	private IdentityMap<PersistentCollection<?>, CollectionEntry> collectionEntries;
+	private InstanceIdentityMap<PersistentCollection<?>, CollectionEntry> collectionEntries;
+
+	// current collection instance id and stack of reusable ones
+	private StandardStack<Integer> reusableCollectionInstanceIds = null;
```

Comment:
I wonder what's the reasoning behind needing to reuse IDs, vs. just using the next int? Performance wrt cached `Integer` instances (<128)? Worries that we'll run out of ints (!)?

Maybe this should be documented as a comment, because I sure won't remember if I run into this code in a few years :)

Reply:
Mainly it's meant to not leave gaps in the underlying `PagedArray` structure: this is a new collection that grows of a constant `PAGE_SIZE` (I went for `32` but we can tune this) as needed to allocate more space. If we don't re-use ids, we risk leaving gaps that increase memory usage and could potentially be numerous (thinking of use cases where entities are removed from the PC regularly before adding more within the same session).

It's not required of course, but it's a pretty simple solution to "compress" the storage of ids.

Reply:
Okay, seems you're going very far into your optimizations... probably way further than I would've, but well, I suppose y'all have your reasons.

Anyway, this definitely deserves some comment(s) in the code, because debugging this stuff will be challenging :)

Reply:
Did you measure the impact of this change on a particular scenario? I would assume that the most common use case that could cause fragmentation is when a persistence context is cleared, in which case the `currentCollectionInstanceId` and `reusableCollectionInstanceIds` are both reset.
So the only case that I can imagine where this would be effective is if collections were detached explicitly or through cascades, which I honestly doubt is happening in real world applications. I really haven't seen a lot of `EntityManager#detach` usage out in the wild, but maybe my observations were within a different bubble.

Reply:
I did not measure it, though that could be certainly arranged. The gaps would cause increased allocations in `PagedArray`, and while certainly `detach` is not used that often it's also about `remove`. I could see use cases where this might happen a lot, and even though they might be the minority, the cost of keeping the ids around is pretty trivial with `StandardStack`.

Reply:
If we can document justified cases that benefit from this optimization I'm totally ok to have it, I just wanted to know what the use cases are and what the impact is, because depending on the use case that benefits, there maybe a different way to handle this :)

Reply:
For now, I've removed the reuse of instance-ids: the most common use case would not benefit much from this, and we can accept a slightly increased memory footprint in case of intensive remove / detach operations (which should not be very frequent) while having a simpler base behavior.

We can always go back to keeping stacks of reusable instance-ids in the future if needed.